feat(api): handling queries responding with something else than JSON

YouriT opened the pull request:
This PR proposes a fix to the issue #4

phillipj commented:
This looks really promising! I would merge if it had a couple of tests
verifying this behavior :) Wouldnt want to break this functionality in the
future..
On tir. 5. mai 2015 at 22.18 Youri <notifications@github.com> wrote:

> YouriT wants to merge 1 commit into phillipj:master from YouriT:master:
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/phillipj/node-plex-api/pull/5
> Commit Summary
>
>    - feat(api): handling queries responding with something else than JSON
>
> File Changes
>
>    - *M* lib/api.js
>    <https://github.com/phillipj/node-plex-api/pull/5/files#diff-0> (17)
>
> Patch Links:
>
>    - https://github.com/phillipj/node-plex-api/pull/5.patch
>    - https://github.com/phillipj/node-plex-api/pull/5.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phillipj/node-plex-api/pull/5>.
>

YouriT commented:
I did some enhancements because I want to have more info on medias. So I added a XML parser because when you add the parameter "includeExtras" to the URL, it responds a mix of JSON / XML which is not working at all.

YouriT commented:
About the tests, the only issue I have is that tests are not passing on my machine, I get the following:
```
lib/api.js: line 12, col 5, Missing "use strict" statement.
lib/api.js: line 31, col 5, Missing "use strict" statement.
lib/api.js: line 35, col 5, Missing "use strict" statement.
lib/api.js: line 39, col 5, Missing "use strict" statement.
lib/api.js: line 47, col 5, Missing "use strict" statement.
lib/api.js: line 82, col 5, Missing "use strict" statement.
lib/api.js: line 92, col 5, Missing "use strict" statement.
lib/api.js: line 141, col 5, Missing "use strict" statement.
lib/api.js: line 154, col 35, 'filterChildrenByCriterias' was used before it was defined.
lib/api.js: line 155, col 5, Missing "use strict" statement.
lib/api.js: line 162, col 29, 'criteriasMatchChild' was used before it was defined.
lib/api.js: line 163, col 5, Missing "use strict" statement.
lib/api.js: line 171, col 29, 'generateRelativeUrl' was used before it was defined.
lib/api.js: line 172, col 5, Missing "use strict" statement.

lib/auth.js: line 12, col 5, Missing "use strict" statement.
lib/auth.js: line 18, col 5, Missing "use strict" statement.
lib/auth.js: line 48, col 5, Missing "use strict" statement.
lib/auth.js: line 56, col 5, Missing "use strict" statement.

lib/uri.js: line 2, col 5, Missing "use strict" statement.
lib/uri.js: line 6, col 5, Missing "use strict" statement.
lib/uri.js: line 17, col 9, Missing "use strict" statement.
lib/uri.js: line 21, col 9, Missing "use strict" statement.
lib/uri.js: line 26, col 5, Missing "use strict" statement.
```

phillipj commented:
Hm, that's strange.. I just checked out your `master` branch and runned:

```bash
$ rm -rf node_modules && npm i && npm test
```

The only style issue I get is the one Travis CI are currently reporting on this pull request, e.g. https://travis-ci.org/phillipj/node-plex-api/jobs/61785999
```bash
> jshint lib/* && jscs lib/* && mocha --reporter list test/*-test.js

Missing line feed at file end at lib/api.js :
   173 |}
   174 |
   175 |module.exports = PlexAPI;
---------------------------------^
```

I have these versions of jshint and jscs installed after a fresh install
```bash
→ npm ls jshint jscs
plex-api@2.0.1 /Users/flipp/Documents/Development/js/node/plex-api
├── jscs@1.13.1 
└── jshint@2.7.0 
```

Could you re-install all dependencies and check if your issues still remain? If so please add your exact versions aswell.

YouriT commented:
I've npm 2.9.0 that's the only difference I see honestly

phillipj commented:
Updated to npm v2.9.0 here aswell, after your last commits there are no style issues reported locally or on Travis - just some broken tests ;) Tell me if I could help on this in some way, looking forward to getting this fix into master.

YouriT commented:
Well I didn't get at all the issue so I added a jshintConfig to the package.json to make it works.

The issue I'm facing is that I need to make pass the test called "query() resource endpoint should return a buffer"

but it seems that nock tries to parse the answer which is a Buffer and it obviously can't be done. I went through the docs of nock and I don't really know how to proceed. Maybe you have an idea ?

By the way I removed the Accept auto-change for includeExtras parameter because I don't really know what it contains more... I just needed checkFiles parameter and this endpoint on the Plex API formats properly the JSON.

Anyway if the endpoint answers XML it's transformed into JSON, it could be helpful one day...

phillipj commented:
Nice @YouriT, I'll have a look at the test and think of some hackery in the test server around the Buffer issue you're having

YouriT commented:
Cool :) thx !

YouriT commented:
Yepa, everything seems fine :) thx !

phillipj commented:
Nice! I'll happily contribute more if you want, just point me in the right direction ;)

phillipj commented:
Yo @YouriT! Any progress on this? I just re-runned the Travis tests which and it all looks good at this 

Anything in particular you are missing of functionality and tests at this moment?

YouriT commented:
Hey ! Everything is fine for me :)

phillipj commented:
Nice! As soon as you squash all these commits into one single commit, I'll merge it
